src/metabase/query_processor/middleware/resolve_referenced.clj

```diff
@@ -0,0 +1,36 @@
+(ns metabase.query-processor.middleware.resolve-referenced
+  (:require [metabase.query-processor.middleware
+             [resolve-fields :as qp.resolve-fields]
+             [resolve-source-table :as qp.resolve-tables]]
+            [schema.core :as s]
+            [toucan.db :as db]))
+
+(defn tags-referenced-cards
+  "Returns Card instances referenced by the given native `query`."
+  [query]
+  (->> (get-in query [:native :template-tags])
+       vals
+       (filter #(= (keyword (:type %)) :card))
+       (map :card)
```

Comment:
since only type=card has a `:card` field, vals+filter+map could be simplified to `(keep (comp :card val)`